src/IO/WriteBufferFromS3.cpp

```diff
@@ -108,7 +111,7 @@ void WriteBufferFromS3::nextImpl()
                 "Cannot write to prefinalized buffer for S3, the file could have been created with PutObjectRequest");
 
     /// Make sense to call to before adding new async task to check if there is an exception
-    task_tracker->waitReady();
+    task_tracker->consumeReady();
```

Comment:
Well. I have a better answer.
In case when `max_tasks_inflight = 0` `add` waits nothing. However we need some point to collect statuses (exception actually since the func returns void). This is that point. We see is any one finished and if so we checks if there is exception.